neonvm/controllers/virtualmachine_controller.go

```diff
@@ -744,55 +744,109 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
-			log.Info("Restarting VM runner pod", "VM.Phase", virtualmachine.Status.Phase, "RestartPolicy", virtualmachine.Spec.RestartPolicy)
-			// get runner to delete
-			found := true
-			vmRunner := &corev1.Pod{}
-			err := r.Get(ctx, types.NamespacedName{Name: virtualmachine.Status.PodName, Namespace: virtualmachine.Namespace}, vmRunner)
-			if apierrors.IsNotFound(err) {
-				found = false
-			} else if err != nil {
-				return err
+			virtualmachine.Status.Phase = phase
+
+			shouldRestart := virtualmachine.Spec.RestartPolicy == vmv1.RestartPolicyAlways ||
+				(virtualmachine.Spec.RestartPolicy == vmv1.RestartPolicyOnFailure && virtualmachine.Status.Phase == vmv1.VmFailed)
+
+			if shouldRestart {
+				log.Info("Restarting VM runner pod", "VM.Phase", virtualmachine.Status.Phase, "RestartPolicy", virtualmachine.Spec.RestartPolicy)
+				virtualmachine.Status.Phase = "" // reset to trigger restart
 			}
-			// delete runner only when VM failed
-			if found && virtualmachine.Status.Phase == vmv1.VmFailed {
-				// delete current runner
-				if err := r.deleteRunnerPodIfEnabled(ctx, virtualmachine, vmRunner); err != nil {
-					return err
+
+			// TODO for RestartPolicyNever: implement TTL or do nothing
+		}
+	default:
+		// do nothing
+	}
+
+	return nil
+}
+
+type runnerStatus string
+
+const (
+	runnerUnknown   runnerStatus = "Unknown"
+	runnerPending   runnerStatus = "Pending"
+	runnerRunning   runnerStatus = "Running"
+	runnerFailed    runnerStatus = "Failed"
+	runnerSucceeded runnerStatus = "Succeeded"
+)
+
+func runnerContainerStatus(pod *corev1.Pod) runnerStatus {
+	switch pod.Status.Phase {
+	case "", corev1.PodPending:
+		return runnerPending
+	case corev1.PodSucceeded:
+		return runnerSucceeded
+	case corev1.PodFailed:
+		return runnerFailed
+	case corev1.PodUnknown:
+		return runnerUnknown
+
+	// With runners with neonvm-container-mgr, the pod phase itself won't be updated if just the
+	// neonvm-runner container exits or completes.
+	// So we need to manually check through the container statuses to check whether they're ok.
+	case corev1.PodRunning:
+		completed := false
+
+		for _, stat := range pod.Status.ContainerStatuses {
+			if stat.State.Terminated != nil {
+				if stat.State.Terminated.ExitCode == 0 {
+					completed = true
+				} else {
+					// exit code non-zero, so the container failed
+					return runnerFailed
 				}
 			}
-			// do cleanup when VM failed OR pod not found (deleted manually?) when VM succeeded
-			if virtualmachine.Status.Phase == vmv1.VmFailed || (!found && virtualmachine.Status.Phase == vmv1.VmSucceeded) {
-				virtualmachine.Cleanup()
-			}
-		case vmv1.RestartPolicyNever:
-			// TODO: implement TTL or do nothing
-		default:
-			// do nothing
+		}
+
+		if completed {
+			return runnerSucceeded
+		} else {
+			return runnerRunning
 		}
 
 	default:
-		// do nothing
+		panic(fmt.Errorf("unknown pod phase: %q", pod.Status.Phase))
 	}
+}
 
-	return nil
+func runnerContainerStopped(pod *corev1.Pod) bool {
```

Comment:
from the function name, i'd expect it to be something like `runnerContainerStatus(pod) != runnerRunning`

Reply:
Changed the name to `runnerStatus` (1da6e055e8c6541325cbf59b5679608bb9dca11c), also to help differentiate from `runnerContainerStopped`. wdyt?

Reply:
Yes, makes sense